core/thread/inc/TPool.h

```diff
@@ -0,0 +1,204 @@
+#ifndef TPOOL_H_
+#define TPOOL_H_
+
+#include "TCollection.h"
+#include "TObjArray.h"
+#include "ROOT/TSeq.h"
+
+template<class subc>
+class TPool {
+public:
+   explicit TPool(){};
+   explicit TPool(size_t nThreads){};
+
+   // // Map
+   // //these late return types allow for a compile-time check of compatibility between function signatures and args,
+   // //and a compile-time check that the argument list implements a front() method (all STL sequence containers have it)
+   template<class F> auto Map(F func, unsigned nTimes) -> std::vector<decltype(func())>;
+   template<class F, class T> auto Map(F func, T &args) -> std::vector < decltype(++(args.begin()), args.end(), func(args.front())) >;
```

Comment:
Deleting the `Map(F func, T &args)` and the `Map(F func, TCollection &args)` overloads and forcing the user to  provide an `std::vector` as the second argument (that means using the `Map(F func, std::vector<T> &args)` overload) for the "collection" use case.

This will simplify the interface and also avoid leaving the input collections in a moved-from state and therefore, unusable.